azure/compute/vm.go

```diff
@@ -0,0 +1,153 @@
+package compute
+
+import (
+	"fmt"
+
+	"github.com/DataDog/test-infra-definitions/azure"
+	"github.com/DataDog/test-infra-definitions/common/utils"
+
+	"github.com/pulumi/pulumi-azure-native-sdk/compute"
+	"github.com/pulumi/pulumi-azure-native-sdk/network"
+	"github.com/pulumi/pulumi-random/sdk/v4/go/random"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+const (
+	imageURNSeparator = ":"
+	adminUsername     = "azureuser"
+)
+
+func NewLinuxInstance(e azure.Environment, name, imageUrn, instanceType string, userData pulumi.StringPtrInput) (*compute.VirtualMachine, *network.PublicIPAddress, *network.NetworkInterface, error) {
+	sshPublicKey, err := utils.GetSSHPublicKey(e.DefaultPublicKeyPath())
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	linuxOsProfile := compute.OSProfileArgs{
+		ComputerName:  pulumi.String(name),
+		AdminUsername: pulumi.String(adminUsername),
+		LinuxConfiguration: compute.LinuxConfigurationArgs{
+			DisablePasswordAuthentication: pulumi.BoolPtr(true),
+			Ssh: compute.SshConfigurationArgs{
+				PublicKeys: compute.SshPublicKeyTypeArray{
+					compute.SshPublicKeyTypeArgs{
+						KeyData: sshPublicKey,
+						Path:    pulumi.StringPtr(fmt.Sprintf("/home/%s/.ssh/authorized_keys", adminUsername)),
+					},
+				},
+			},
+		},
+		CustomData: userData,
+	}
+
+	return newVMInstance(e, name, imageUrn, instanceType, linuxOsProfile)
+}
+
+func NewWindowsInstance(e azure.Environment, name, imageUrn, instanceType string, userData, firstLogonCommand pulumi.StringPtrInput) (*compute.VirtualMachine, *network.PublicIPAddress, *network.NetworkInterface, pulumi.StringOutput, error) {
+	windowsAdminPassword, err := random.NewRandomPassword(e.Ctx, e.Namer.ResourceName(name, "admin-password"), &random.RandomPasswordArgs{
+		Length:  pulumi.Int(20),
+		Special: pulumi.Bool(true),
+	})
+	if err != nil {
+		return nil, nil, nil, pulumi.StringOutput{}, err
+	}
+
+	windowsOsProfile := compute.OSProfileArgs{
+		ComputerName:  pulumi.String(name),
+		AdminUsername: pulumi.String(adminUsername),
+		AdminPassword: windowsAdminPassword.Result,
+		CustomData:    userData,
+	}
+
+	if firstLogonCommand != nil {
+		windowsOsProfile.WindowsConfiguration = compute.WindowsConfigurationArgs{
+			AdditionalUnattendContent: compute.AdditionalUnattendContentArray{
+				compute.AdditionalUnattendContentArgs{
+					ComponentName: compute.ComponentNames_Microsoft_Windows_Shell_Setup,
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**❓ question**
Is this to allow ssh connections?

Reply:
No, it's a way to run script on first-login. I **guess** it could be used to setup SSH, but I haven't verified that yet.